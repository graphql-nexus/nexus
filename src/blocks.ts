export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
