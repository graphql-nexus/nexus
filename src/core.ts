// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
