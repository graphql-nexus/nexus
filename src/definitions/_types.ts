import {
  GraphQLCompositeType,
  GraphQLEnumType,
  GraphQLFieldConfig,
  GraphQLInputObjectType,
  GraphQLInputObjectTypeConfig,
  GraphQLInterfaceType,
  GraphQLInterfaceTypeConfig,
  GraphQLLeafType,
  GraphQLObjectType,
  GraphQLObjectTypeConfig,
  GraphQLScalarType,
  GraphQLSchema,
  GraphQLUnionType,
} from 'graphql'
import {
  NexusFieldExtension,
  NexusInputObjectTypeExtension,
  NexusInterfaceTypeExtension,
  NexusObjectTypeExtension,
  NexusSchemaExtension,
} from '../extensions'
import * as AbstractTypes from '../typegenAbstractTypes'
import { RequiredDeeply } from '../typeHelpersInternal'

export { AbstractTypes }

export type Maybe<T> = T | null

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type BaseScalars = 'String' | 'Int' | 'Float' | 'ID' | 'Boolean'

export enum NexusTypes {
  Arg = 'Arg',
  DynamicInput = 'DynamicInput',
  DynamicOutputMethod = 'DynamicOutputMethod',
  DynamicOutputProperty = 'DynamicOutputProperty',
  Enum = 'Enum',
  ExtendInputObject = 'ExtendInputObject',
  ExtendObject = 'ExtendObject',
  InputField = 'InputField',
  InputObject = 'InputObject',
  Interface = 'Interface',
  List = 'List',
  NonNull = 'NonNull',
  Null = 'Null',
  Object = 'Object',
  OutputField = 'OutputField',
  Plugin = 'Plugin',
  PrintedGenTyping = 'PrintedGenTyping',
  PrintedGenTypingImport = 'PrintedGenTypingImport',
  Scalar = 'Scalar',
  Union = 'Union',
}

export interface DeprecationInfo {
  /**
   * Reason for the deprecation.
   */
  reason: string
  /**
   * Date | YYYY-MM-DD formatted date of when this field
   * became deprecated.
   */
  startDate?: string | Date
  /**
   * Field or usage that replaces the deprecated field.
   */
  supersededBy?: string
}

export interface NonNullConfig {
  /**
   * Whether output fields are non-null by default.
   *
   * type Example {
   *   field: String!
   *   otherField: [String!]!
   * }
   *
   * @default false
   */
  output?: boolean
  /**
   * Whether input fields (field arguments, input type members)
   * are non-null by default.
   *
   * input Example {
   *   field: String
   *   something: [String]
   * }
   *
   * @default false
   */
  input?: boolean
}

export type GraphQLPossibleOutputs = GraphQLCompositeType | GraphQLLeafType

export type GraphQLPossibleInputs = GraphQLInputObjectType | GraphQLLeafType

export const NexusWrappedSymbol = Symbol.for('@nexus/wrapped')

export function withNexusSymbol(obj: Function, nexusType: NexusTypes) {
  obj.prototype[NexusWrappedSymbol] = nexusType
}

export interface AsyncIterator<T> {
  next(value?: any): Promise<IteratorResult<T>>
  return?(value?: any): Promise<IteratorResult<T>>
  throw?(e?: any): Promise<IteratorResult<T>>
}

export type RootTypingDef = string | TypingImport

export type RootTypings = Record<string, string | TypingImport>

export interface TypingImport {
  /**
   * Node module, or absolute path to import the type from.
   */
  path: string
  /**
   * Name of the type we want to reference in the `path`
   */
  name: string
  /**
   * Name we want the imported type to be referenced as
   */
  alias?: string
}

export interface MissingType {
  fromObject: boolean
}

export type GraphQLNamedOutputType =
  | GraphQLScalarType
  | GraphQLObjectType
  | GraphQLInterfaceType
  | GraphQLUnionType
  | GraphQLEnumType

export type GraphQLNamedInputType = GraphQLScalarType | GraphQLInputObjectType | GraphQLEnumType

type WithExt<T extends { extensions?: any }, Ext> = Omit<T, 'extensions'> & {
  extensions?: Maybe<{ nexus?: Ext }>
}

export type NexusGraphQLFieldConfig = WithExt<GraphQLFieldConfig<any, any>, NexusFieldExtension> & {
  name: string
}

export type NexusGraphQLObjectTypeConfig = WithExt<
  GraphQLObjectTypeConfig<any, any>,
  NexusObjectTypeExtension
>

export type NexusGraphQLInputObjectTypeConfig = WithExt<
  GraphQLInputObjectTypeConfig,
  NexusInputObjectTypeExtension
>

export type NexusGraphQLInterfaceTypeConfig = WithExt<
  GraphQLInterfaceTypeConfig<any, any>,
  NexusInterfaceTypeExtension
> & { interfaces: () => GraphQLInterfaceType[] }

export type NexusGraphQLSchema = Omit<GraphQLSchema, 'extensions'> & {
  extensions: { nexus: NexusSchemaExtension }
}

export type NexusFeaturesInput = {
  /**
   * Toggle runtime checks for correct implementation of abstract types. This is
   * a redundant check Nexus makes over the existing static typings it provides.
   *
   * @remarks
   *
   * This is useful for beginners because Nexus can give clear concise error
   * messages unlike the static type errors.
   *
   * Note that if you enable the "abstractTypeStrategies.__typename" feature
   * then this feature will be automatically disabled. For why this is, see that
   * features' remarks.
   */
  abstractTypeRuntimeChecks?: boolean
  /**
   * Toggle abstract-type strategies. For more detail about this feature please refer to to the [abstract types guide](https://nxs.li/guides/abstract-types).
   *
   * If you plan on enabling multiple strategies and you've never done so then please [read the guide about using multiple strategies](https://nxs.li/guides/abstract-types/using-multiple-strategies) as there are a few quirks to be aware of.
   *
   * @default
   *
   * {
   *    resolveType: true,
   *    __typename: false
   *    isTypeOf: false,
   * }
   */
  abstractTypeStrategies?: {
    /**
     * The Modular abstract type strategy. Every member object of an abstract
     * type (union members or interface implementors) will generally be required
     * to implement isTypeOf method. Nexus will not require it in cases where it
     * detects you have implemented another strategy. For more detail see the
     * guide for the [Modular Abstract Type Strategy](https://nxs.li/guides/abstract-types/modular-strategy).
     */
    isTypeOf?: boolean
    /**
     * The Centralized abstract type strategy. Every abstract type (union or
     * interface) will generally be required to implement its resolveType
     * method. Nexus will not require it in cases where it detects you have
     * implemented another strategy. For more detail see the guide for the [Central Abstract Type Strategy](https://nxs.li/guides/abstract-types/centralized-strategy).
     */
    resolveType?: boolean
    /**
     * The Discriminant Model Field strategy. In this mode the resolvers of fields typed as
     * abstract types will be required to include "__typename" field in the
     * returned data. For more detail see the guide for the [Discriminant Model Field Strategy](https://nxs.li/guides/abstract-types/discriminant-model-field-strategy).
     *
     * @warning
     *
     * When this strategy is enabled in conjunction with other strategies the "abstractTypeRuntimeChecks" feature will
     * automatically be disabled. This is because it is not practical at runtime
     * to find out if resolvers will return objects that include the
     * "__typename" field. This trade-off can be acceptable since the runtime checks are a redundant safety
     * measure over the static typing. So as long as you are not ignoring static errors related to Nexus' abstract type type checks then you then you should still have a safe implementation.
     *
     * Furthermore another effect is that statically the other strategies will not appear to be _required_, but instead _optional_, while only this one will appear required. However, upon implementing any of the other strategies, this one will not longer be required. This quirk is explained in the guide section about [using multiple strategies](https://nxs.li/guides/abstract-types/using-multiple-strategies).
     *
     */
    __typename?: boolean
  }
}

export type NexusFeatures = RequiredDeeply<NexusFeaturesInput>
