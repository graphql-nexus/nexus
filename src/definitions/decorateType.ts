import type { GraphQLNamedType } from 'graphql'
import type { SourceTypingDef } from './_types'

export interface TypeExtensionConfig {
  asNexusMethod?: string
  sourceType?: SourceTypingDef
}

export function decorateType<T extends GraphQLNamedType>(type: T, config: TypeExtensionConfig): T {
  type.extensions = {
    ...type.extensions,
    nexus: {
      ...Object(type.extensions?.nexus),
      ...config,
    },
  }
  return type
}
