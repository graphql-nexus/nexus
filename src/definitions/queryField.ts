import { FieldOutConfig, OutputDefinitionBlock } from '../core'
import { extendType, NexusExtendTypeDef } from './extendType'

export type QueryFieldConfig<FieldName extends string> =
  | FieldOutConfig<'Query', FieldName>
  | (() => FieldOutConfig<'Query', FieldName>)

export function queryField(fieldFn: (t: OutputDefinitionBlock<'Query'>) => void): NexusExtendTypeDef<'Query'>

export function queryField<FieldName extends string>(
  fieldName: FieldName,
  config: QueryFieldConfig<FieldName>
): NexusExtendTypeDef<'Query'>

export function queryField(...args: any[]) {
  return extendType({
    type: 'Query',
    definition(t) {
      if (typeof args[0] === 'function') {
        return args[0](t)
      }
      const [fieldName, config] = args as [string, QueryFieldConfig<any>]
      const finalConfig = typeof config === 'function' ? config() : config
      t.field(fieldName, finalConfig)
    },
  })
}
