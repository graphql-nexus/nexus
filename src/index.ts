import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
