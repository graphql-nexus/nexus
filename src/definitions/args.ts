import type { GraphQLScalarTypeConfig } from 'graphql'
import type { AllInputTypes, GetGen2 } from '../typegenTypeHelpers'
import type { AllNexusArgsDefs, AllNexusInputTypeDefs } from './wrapping'
import { Maybe, NexusTypes, withNexusSymbol } from './_types'

export type ArgsRecord = Record<string, AllNexusArgsDefs>

export type CommonArgConfig = {
  /**
   * [GraphQL 2018 Spec](https://spec.graphql.org/June2018/#sec-Descriptions)
   *
   * The description for this argument.
   *
   * Various GraphQL tools will make use of this information but it has zero runtime impact. The value given
   * here will also be included as heredocs in the generated GraphQL SDL file.
   *
   * Default :: By default there will be no description
   *
   * @example
   *   export const Query = queryType({
   *     definition(t) {
   *       t.list.int('connectionCount', {
   *         args: {
   *           lastSeconds: intArg({
   *             description: 'Limit count to this number of past seconds from now.',
   *           }),
   *         },
   *       })
   *     },
   *   })
   *
   *   // GraphQL SDL
   *   // -----------
   *   //
   *   // type Query {
   *   //   connectionCount(
   *   //   """Limit count to this number of past seconds from now."""
   *   //    lastSeconds: Int
   *   //   ): [Int]
   *   // }
   */
  description?: Maybe<string>

  /**
   * Data that will be added to the arg-level [extensions field on the graphql-js type def
   * instances](https://github.com/graphql/graphql-js/issues/1527) resulting from makeSchema. Useful for some
   * graphql-js based tools which rely on looking for special data here.
   */
  extensions?: GraphQLScalarTypeConfig<any, any>['extensions']
} & NexusGenPluginArgConfig

export interface ScalarArgConfig<T> extends CommonArgConfig {
  /**
   * The default value for this argument when ***none*** is given by the client.
   *
   * Note that *null* is still considered something meaning if the client gives an explicit null that will
   * prevent the default from activating. This is why the type of an arg with a default value in the resolver
   * includes "undefined | null".
   *
   * @example
   *   intArg({ default: 42 })
   */
  default?: T
}

export type NexusArgConfigType<T extends string> = T | AllNexusInputTypeDefs<T>

export interface NexusAsArgConfig<T extends string> extends CommonArgConfig {
  /**
   * Sets the default value for this argument, should match the type of the argument
   *
   * @example
   *   intArg({ default: 42 })
   */
  default?: GetGen2<'inputTypeShapes', T>
}

export interface NexusArgConfig<T extends string> extends NexusAsArgConfig<T> {
  /**
   * [GraphQL 2018 Spec](https://spec.graphql.org/June2018/#sec-Types)
   *
   * The type that this argument should be.
   *
   * Argument types may be scalars or other input types (input objects, enums) in your schema and modified
   * with non-null and list types.
   *
   * Types may be expressed in one of three ways:
   *
   * 1. As string literals matching the name of a builtin scalar. 2. As string literals matching the name of
   * another type. Thanks to [Nexus' reflection
   *     system](https://nxs.li/guides/reflection) this is typesafe and autocompletable. 3. As references to other
   * enums or input object type definitions.
   *
   * Type modifier helpers like list() may also be used and in turn accept one of the three methods listed above.
   *
   * @example
   *   arg({ type: 'UserCreateInput' })
   *
   * @example
   *   arg({ type: nonNull(list(nonNull('UserCreateInput'))) })
   *
   * @example
   *   arg({ type: UserCreateInput })
   *
   * @example
   *   arg({ type: nonNull(list(nonNull(UserCreateInput))) })
   *
   * @example
   *   arg({ type: 'String' })
   */
  type: NexusArgConfigType<T>
}

export interface NexusFinalArgConfig extends NexusArgConfig<any> {
  configFor: 'arg'
  argName: string
  fieldName: string
  parentType: string
}

export class NexusArgDef<TypeName extends AllInputTypes> {
  constructor(readonly name: TypeName, protected config: NexusArgConfig<any>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(NexusArgDef, NexusTypes.Arg)

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define an argument. Arguments can be used with the args config of any field.
 *
 * @example
 *   export const Mutation = mutationType({
 *     definition(t) {
 *       t.field('createDraft', {
 *         type: 'Post',
 *         args: {
 *           title: arg({
 *             type: 'String',
 *             default: 'Untitled',
 *             description: 'The title of this draft post.',
 *           }),
 *           body: nonNull(
 *             arg({
 *               type: 'String',
 *               description: 'The content of this draft post.',
 *             })
 *           ),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its type and description. See jsdoc on each config field
 *     for details.
 */
export function arg<T extends string>(config: NexusArgConfig<T>) {
  if (!config.type) {
    throw new Error('You must provide a "type" for the arg()')
  }
  return new NexusArgDef(typeof config.type === 'string' ? config.type : (config.type as any).name, config)
}

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define a string argument. Arguments can be used with the args config of any field.
 *
 * This is a shorthand equivalent to:
 *
 * `arg({ type: 'String', ... })`
 *
 * @example
 *   export const Mutation = mutationType({
 *     definition(t) {
 *       t.field('createDraft', {
 *         type: 'Post',
 *         args: {
 *           title: stringArg(),
 *           body: stringArg(),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its description. See jsdoc on each config field for details.
 */
export function stringArg(config?: ScalarArgConfig<string>) {
  return arg({ type: 'String', ...config })
}

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define a string argument. Arguments can be used with the args config of any field.
 *
 * This is a shorthand equivalent to:
 *
 * `arg({ type: 'Int', ... })`
 *
 * @example
 *   export const Query = queryType({
 *     definition(t) {
 *       t.field('search', {
 *         type: 'SearchResult',
 *         args: {
 *           maxResults: intArg(),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its description. See jsdoc on each config field for details.
 */
export function intArg(config?: ScalarArgConfig<number>) {
  return arg({ type: 'Int', ...config })
}

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define a string argument. Arguments can be used with the args config of any field.
 *
 * This is a shorthand equivalent to:
 *
 * `arg({ type: 'Float', ... })`
 *
 * @example
 *   export const Query = queryType({
 *     definition(t) {
 *       t.field('search', {
 *         type: 'SearchResult',
 *         args: {
 *           ratingAbove: floatArg(),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its description. See jsdoc on each config field for details.
 */
export function floatArg(config?: ScalarArgConfig<number>) {
  return arg({ type: 'Float', ...config })
}

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define a string argument. Arguments can be used with the args config of any field.
 *
 * This is a shorthand equivalent to:
 *
 * `arg({ type: 'ID', ... })`
 *
 * @example
 *   export const Query = queryType({
 *     definition(t) {
 *       t.field('user', {
 *         type: 'User',
 *         args: {
 *           id: idArg(),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its description. See jsdoc on each config field for details.
 */
export function idArg(config?: ScalarArgConfig<string>) {
  return arg({ type: 'ID', ...config })
}

/**
 * [API Docs](https://nexusjs.org/docs/api/args) | [GraphQL.org
 * Docs](https://graphql.github.io/learn/schema/#arguments) | [GraphQL 2018
 * Spec](https://spec.graphql.org/June2018/#sec-Language.Arguments)
 *
 * Define a string argument. Arguments can be used with the args config of any field.
 *
 * This is a shorthand equivalent to:
 *
 * `arg({ type: 'Boolean', ... })`
 *
 * @example
 *   export const Query = queryType({
 *     definition(t) {
 *       t.list.field('users', {
 *         type: 'User',
 *         args: {
 *           active: booleanArg(),
 *         },
 *         // ...
 *       })
 *     },
 *   })
 *
 * @param config Configuration for the argument like its description. See jsdoc on each config field for details.
 */
export function booleanArg(config?: ScalarArgConfig<boolean>) {
  return arg({ type: 'Boolean', ...config })
}
