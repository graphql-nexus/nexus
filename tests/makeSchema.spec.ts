import { printSchema } from 'graphql'
import os from 'os'
import path from 'path'
import { generateSchema, makeSchema } from '../src/builder'
import { queryField } from '../src/definitions/queryField'

describe('makeSchema', () => {
  describe('shouldExitAfterGenerateArtifacts', () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('exits with 0 code after successful write', (done) => {
      const logSpy = jest.spyOn(console, 'log').mockImplementationOnce(() => {})
      jest.spyOn(process, 'exit').mockImplementationOnce((code) => {
        expect(code).toEqual(0)
        expect(logSpy.mock.calls[0][0]).toContain('Generated Artifacts:')
        return done() as never
      })
      makeSchema({
        types: [
          queryField('someField', {
            type: 'String',
            resolve: () => 'Test',
          }),
        ],
        outputs: {
          typegen: path.join(os.tmpdir(), '/file.ts'),
          schema: path.join(os.tmpdir(), '/schema.graphql'),
        },
        shouldGenerateArtifacts: true,
        shouldExitAfterGenerateArtifacts: true,
      })
    })

    // using jest.spyOn on process.exit doesn't not work on Windows
    if (process.platform !== 'win32') {
      it('exits with 1 code and logs error after failure', (done) => {
        const errSpy = jest.spyOn(console, 'error').mockImplementationOnce(() => {})
        jest.spyOn(process, 'exit').mockImplementationOnce((code) => {
          expect(code).toEqual(1)
          expect(errSpy.mock.calls[0][0].message).toEqual(
            `ENOTDIR: not a directory, open '/dev/null/schema.graphql'`
          )
          expect(errSpy.mock.calls.length).toEqual(1)
          return done() as never
        })
        makeSchema({
          types: [
            queryField('someField', {
              type: 'String',
              resolve: () => 'Test',
            }),
          ],
          outputs: {
            typegen: path.normalize(`/dev/null/file.ts`),
            schema: path.normalize(`/dev/null/schema.graphql`),
          },
          shouldGenerateArtifacts: true,
          shouldExitAfterGenerateArtifacts: true,
        })
      })
    }

    it('accepts a customPrintSchemaFn', async () => {
      const { schemaTypes } = await generateSchema.withArtifacts(
        {
          types: [
            queryField('ok', {
              description: 'Example boolean field',
              type: 'Boolean',
            }),
          ],
          outputs: {},
          customPrintSchemaFn: (schema) => {
            return printSchema(schema, { commentDescriptions: true })
          },
        },
        false
      )
      expect(schemaTypes).toMatchSnapshot()
    })
  })

  describe('contextType', () => {
    it('can specify contextType as a typing import', async () => {
      const { tsTypes } = await generateSchema.withArtifacts(
        {
          outputs: false,
          types: [
            queryField('ok', {
              description: 'Example boolean field',
              type: 'Boolean',
            }),
          ],
          shouldGenerateArtifacts: true,
          typegenAutoConfig: {
            sources: [],
            contextType: {
              path: 'graphql',
              name: 'GraphQLInputFieldConfigMap',
            },
          },
        },
        path.normalize(`/dev/null/file.ts`)
      )
      expect(tsTypes).toContain(`import { GraphQLInputFieldConfigMap } from "graphql"`)
      expect(tsTypes).toContain(`context: GraphQLInputFieldConfigMap`)
    })
  })
})
