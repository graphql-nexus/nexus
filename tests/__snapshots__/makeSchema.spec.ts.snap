// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeSchema shouldExitAfterGenerateArtifacts accepts a customPrintSchemaFn 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


type Query {
  # Example boolean field
  ok: Boolean
}
"
`;
